### PR TITLE
feat: Split oversized reports into multiple comments

### LINKED ISSUE #504: Split comments instead of truncating when exceeding VCS comment size limits
## Problem

When kubechecks produces a report that exceeds the VCS comment size limit (GitHub: 64 KB, GitLab: 1 MB), the comment is truncated.

This results in incomplete reports and effectively renders kubechecks unusable for pull requests that touch numerous apps or resources with sizeable diffs (e.g., CRDs, bulk app removals). Several users and organisations have flagged this as a blocking issue, with some resorting to maintaining custom builds to work around it.

This has been a long-standing feature request with multiple community contributions (#299, #301, #421, #436) - all of which were eventually closed by the stale bot without being merged. Similar tools (e.g., Atlantis) already handle this gracefully.

It would be lovely to finally get this sorted.

I can work on this as a matter of priority if you like, so let's try to get all the requirements together.

### Configuration

| Environment Var | Type | Default | Description |
|---|---|---|---|
| `KUBECHECKS_MAX_COMMENTS_PER_CHECK` | `int` | `0` (unlimited) | Hard cap on the number of comments posted per check run. `0` means no cap. Acts as an operational guardrail against runaway posting. When the cap is reached, the final comment is truncated with a note indicating that the full output exceeded the maximum number of comments. |

### Comment Splitting Logic

Structural splitting:

- Split at app-level boundaries (`<details>` blocks), keeping each app's section intact within a single comment wherever possible.
- If a single app's output exceeds the VCS limit on its own, fall back to byte-level splitting within that app's section.

Byte-level fallback:

- Close any open markdown constructs (code blocks, `<details>` tags) at the end of each chunk.
- Re-open them at the start of the next chunk.
- Each chunk must be valid, renderable markdown in its own right.

Chunk decoration:

- Every chunk carries the header: `# Kubechecks {identifier} Report (Part N of M)`. The identifier MUST be present in all chunks so that multi-instance deployments sharing the same token correctly prune or hide only their own comments (per the [review on #301](https://github.com/zapier/kubechecks/pull/301#discussion_r2028616667) and #370).
- Non-first chunks include: `Continued from previous comment.`
- Non-last chunks include: `Continued in next comment.`
- The footer (commit SHA, duration, pod info) appears only on the final chunk.

### VCS Client Changes

- `UpdateMessage` signature: Accept `[]string` (chunks) rather than a single `string`. The first chunk edits the existing comment; subsequent chunks are posted as new comments.
- `PostMessage`: `BuildComment` returns `[]string` and `PostMessage` posts them sequentially.
- Backoff and retry: Exponential backoff on VCS API calls during multi-comment posting to respect rate limits.
- GitHub and GitLab: The splitting logic itself is VCS-agnostic. Only `MaxCommentLength` and the underlying API calls differ per provider.

### Clean Up the "kubechecks running" Placeholder

When posting the final report:

- Find and delete the `:hourglass: kubechecks running ...` placeholder comment rather than editing it in place.
- Post the report as fresh comment(s).

This sidesteps the rather awkward in-place edit of a placeholder and simplifies the split-comment flow.

### TidyOutdatedComments

`pruneOldComments` and `hideOutdatedMessages` must account for multi-comment runs:

- Identify all comments belonging to the same kubechecks instance and check run via the identifier in the header.
- Prune or hide the entire group together.

### Memory Considerations

A prior attempt (#436) surfaced excessive memory usage during splitting. The implementation should:
- Avoid materialising the entire split output in memory at once where feasible.
- Favour `strings.Builder` for chunk assembly.
- Profile memory usage with large reports (100 KB+) as part of testing.

### Testing

- Unit tests for the splitting function covering both structural and byte-level paths.
- Unit tests for chunk decoration (identifier, part numbering, markdown validity).
- Manual verification for both GitHub and GitLab (Not sure if integration tests are feasible).
- Edge cases: single app exceeding the limit, report landing exactly at the limit, empty report, single-app report, and the max-comments cap being reached.

## Out of Scope

- Offloading full output to external storage (S3, etc.).
- Per-resource-level splitting (GitHub's 64 KB limit means single-resource diffs can span multiple comments regardless).

## References

- #299, #301, #421, #436 -- prior issues and community attempts.
- #370 -- identifier feature (merged). Identifier must appear in all split chunks.
- [Review comment on #301 requiring the identifier in continued messages](https://github.com/zapier/kubechecks/pull/301#discussion_r2028616667).

Cc: @MeNsaaH, @Greyeye, @djeebus

### PR DESCRIPTION
**Pull request by vovinacci**
A report bigger than the comment limit (GitHub 64 KiB, GitLab 1 MB) was cut at the limit, which lost the end of the report and left us with no hint that anything is missing (and usually broke the markdown too).

At times that heavily confused us and we had Atlantis as a good example of how it works nicely. So, now it is posted as several comments.

Behaviour for a report that fits in one comment does not change: the output is byte-identical to main (there is a test pinning that), and the placeholder comment is edited in place. Only the overflow is posted as new comments, each headed `Kubechecks <identifier> Report (Part N of M)`. Every run posts its own comments; comments from earlier runs, including all parts of a split report, are hidden or deleted by TidyOutdatedComments as before.

How it splits: whole app sections first, then groups of checks, and only then the details of a single check, between lines. Fences and `<details>` open at a cut are closed and reopened in the next part, so every comment renders on its own. The comment at the top of `pkg/msg/split.go` walks through it.

Also in here:
- `KUBECHECKS_MAX_COMMENTS_PER_CHECK`, 1 to 999, default 999 (part numbers are budgeted for three digits). We see 20-30 comments on large changes in production without rate-limit trouble, a low default would truncate those.
- `vcs.Client.UpdateMessage` takes the PR, the note ID and the chunks; `MaxCommentLength()` is new; `BuildComment` takes an options struct.
- GitHub: a failed create is retried on rate limits only, a repeated create would post the chunk twice.
- The commit status is now set even if posting the comment fails. It used to stay pending although all checks had run.

Notes:
- Overflow comments are new comments, so subscribers get a notification for each of them.
- If posting fails part-way, the comments already posted stay, the last of them still saying "Continued in next comment".
- GitLab: `checkReturnForBackoff` treats every error except 429 as permanent. That was one call per report and is now one per comment, so a transient 5xx is more likely to cut a long report short. Happy to follow up separately.

Easiest order to review: `pkg/vcs/types.go` and the two clients, then `pkg/msg` (`split.go` first), then `github_client/retry.go`, which stands on its own.

Resolves #504.

### CODE REVIEW

**Comment by andridzi** [comment]
already tested, works perfectly!

**Comment by vovinacci** [comment]
@djeebus @Greyeye Small heads-up to review once you have some spare time.
Works for me and @andridzi like a charm.

**Comment by vovinacci** [comment]
@Greyeye @djeebus - a gentle nudge on this one. It continues to run without issue for us, and for @andridzi as well.

There has been no feedback so far, so I would like to ask for a decision either way.

To be explicit: "no, we do not want this" is a perfectly acceptable answer. We run it from a fork and can carry on doing so. I would simply prefer a clear decision to leaving the PR open indefinitely.

If the approach is the concern rather than the feature itself, I am happy to rework it - please point me at what you would like changed.

Many thanks.

File: pkg/msg/split.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
+package msg
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/zapier/kubechecks/pkg"
+)
```
[the remark this answers is not in the document]
Comment (vovinacci) [reply pkg/msg/split.go:10]: `slices.Clone` lives in `message.go`, which imports it. `split.go` does not use it.

File: pkg/msg/message.go [hunk 2]
```diff
@@ -149,75 +150,337 @@ func (m *Message) buildFooter(
...
+		MaxChunks:  opts.MaxComments,
+		Identifier: opts.Identifier,
+		Footer:     m.buildFooter(opts.Start, opts.CommitSHA, opts.LabelFilter, opts.ShowDebugInfo, opts.AppsChecked, opts.TotalChecked),
+	}
+
+	var allSections []string
+	for _, appName := range names {
+		if m.isDeleted(appName) {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vovinacci) [reply pkg/msg/message.go]: Two copies of the report at the peak, yes, however that's the original behaviour.

`BuildComment` grew a `strings.Builder` to the full report and `UpdateMessage` then sliced it to 64 KiB (GitHub) or 1 MB (GitLab).

The cap does not add memory. On paper, 999 comments on GitLab is 1 GB of output, however that report needs 1 GB of check output already held in the `Message`, which would have been just as fatal at any time before this PR.

Given the nature of this PR, I didn't have any intentions to change this behaviour as this is itself quite a change (and I'm happy to implement it as a separate effort, @Greyeye).
What I did add is that rendering now stops once the comment cap is reached, so with a low cap the transient is bounded by the cap rather than by the input.

File: pkg/msg/split.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// sectionBudget is the space left for app sections in a chunk. We do not know
+// up front which chunk ends up last, so every chunk is budgeted for the worse
+// of the two tails.
+func (cfg chunkConfig) sectionBudget() int {
+	head := len(chunkHeader(cfg.Identifier, pkg.MaxCommentsPerCheck, pkg.MaxCommentsPerCheck)) + len(continuedFrom)
+	tail := max(len(continuedIn), len(footerSeparator)+len(cfg.Footer)+len(truncatedNote))
+
+	return cfg.MaxLength - head - tail
```
[the remark this answers is not in the document]
Comment (vovinacci) [reply pkg/msg/split.go]: That's a good one.

Every comment was budgeted for the multi-part header and the notes, so a report within ~200 bytes of the limit was split in two although it fitted in a single comment.

Fixed in e23f623 along with the test for a report of exactly the limit.
